File: Task2/apz-pzpi-21-11-milennyi-ivan-task2/src/db/services/shearing_log_service.rs
use std::sync::Arc;
use async_trait::async_trait;
use sqlx::{MySql, Pool, query, query_as};
use crate::db::service_error::ServiceError;
use crate::db::traits::{Service, ShearingLogManage};
use crate::models::ShearingLog;
use crate::view_models::ShearingLogVM;

pub(crate) struct ShearingLogService<T> {
    pool: Arc<T>,
}
#[async_trait]
impl Service<Pool<MySql>> for ShearingLogService<Pool<MySql>> {
    type Model = ShearingLog;
    type Error = ServiceError;
    type ViewModel = ShearingLogVM;
    fn new(pool: Arc<Pool<MySql>>) -> Self {
        ShearingLogService { pool }
    }

    async fn create(&self, item: Self::Model) -> Result<Self::Model, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            INSERT INTO ShearingLogs (sheep_id, timestamp, wool_amount)
            VALUES (?, ?, ?)
            RETURNING id, sheep_id, timestamp, wool_amount
            "#
        )
        .bind(item.sheep_id())
        .bind(item.timestamp())
        .bind(item.wool_amount())
        .fetch_one(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }

    async fn delete(&self, item_id: u64) -> Result<(), Self::Error> {
        query(
            r#"
            DELETE FROM ShearingLogs
            WHERE id = ?
            "#
        )
        .bind(item_id)
        .execute(&*self.pool).await
        .map(|_| ()).map_err(|error| ServiceError::DatabaseError(error))
    }

    async fn update(&self, _item: Self::Model) -> Result<Self::Model, Self::Error> {
        Err(ServiceError::ForbiddenError)
    }

    async fn get_all(&self) -> Result<Vec<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM ShearingLogs
            "#
        )
        .fetch_all(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }

    async fn get_by_id(&self, id: u64) -> Result<Option<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM ShearingLogs
            WHERE id = ?
            "#
        )
       .bind(id)
       .fetch_optional(&*self.pool).await
       .map_err(|error| ServiceError::DatabaseError(error))
    }
}
#[async_trait]
impl ShearingLogManage<Pool<MySql>> for ShearingLogService<Pool<MySql>>{
    async fn get_all_vms_by_sheep_id(&self, id: u64) -> Result<Vec<Self::ViewModel>, Self::Error> {
        todo!()
    }
}

File: Task4/apz-pzpi-21-11-milennyi-ivan-task4/src/api/services/ShearingLogService.ts
import axios from "axios";
import { ShearingLog } from "../../models/ShearingLog";
import { API_URL } from "../../utils/config";
import { ShearingLogVM } from "../../viewModels/ShearingLogVM";
import { IShearingLogService } from "../interfaces/IShearingLogService";

export class ShearingLogService implements IShearingLogService{
    private static SHEARING_LOG_URLS = {
        GET_ALL: `${API_URL}/shearing-log`,
        GET_BY_ID: (id: number) => `${API_URL}/shearing-log/${id}`,
        CREATE: `${API_URL}/shearing-log/create`,
        DELETE: (id: number) => `${API_URL}/shearing-log/delete/${id}`,
        GET_ALL_VMS_BY_SHEEP_ID: (id: number) =>  `${API_URL}/shearing-log/sheep/${id}`
    }

    async create(item: ShearingLog): Promise<ShearingLog> {
        const response = await axios.post<ShearingLog>(ShearingLogService.SHEARING_LOG_URLS.CREATE, item);
        return response.data;
    }
    async delete(itemId: number): Promise<void> {
        await axios.delete(ShearingLogService.SHEARING_LOG_URLS.DELETE(itemId));
    }
    async update(_item: ShearingLog): Promise<ShearingLog> {
        throw new Error("Method is forbidden.");
    }
    async getAll(): Promise<ShearingLog[]> {
        const response = await axios.get<ShearingLog[]>(ShearingLogService.SHEARING_LOG_URLS.GET_ALL);
        return response.data;
    }
    async getById(id: number): Promise<ShearingLog | null> {
        const response = await axios.get<ShearingLog | null>(ShearingLogService.SHEARING_LOG_URLS.GET_BY_ID(id));
        return response.data;
    }
    async getAllVMsBySheepId(id: number): Promise<ShearingLogVM[]> {
        const response = await axios.get<ShearingLogVM[]>(ShearingLogService.SHEARING_LOG_URLS.GET_ALL_VMS_BY_SHEEP_ID(id));
        return response.data;
    }
}